demos/node-classification/ppnp/appnp-reddit.ipynb

```diff
@@ -0,0 +1,415 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "55334/55334 [==============================] - 1s 23us/sample - loss: 1.1373 - acc: 0.7331\n",
+      "\n",
+      "Test Set Metrics:\n",
+      "\tloss: 1.1373\n",
+      "\tacc: 0.7331\n"
+     ]
+    }
+   ],
+   "source": [
+    "test_metrics = fully_connected_model.evaluate(test_data, test_targets)\n",
+    "print(\"\\nTest Set Metrics:\")\n",
+    "for name, val in zip(fully_connected_model.metrics_names, test_metrics):\n",
+    "    print(\"\\t{}: {:0.4f}\".format(name, val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## APPNP Propagation\n",
+    "\n",
+    "We now create an APPNP model and propagate the MLP. No further training is happening in this step.  We then test the propagated model on the test data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "appnp = APPNP(\n",
+    "    layer_sizes=[train_targets.shape[-1]],\n",
+    "    activations=[\"relu\"],\n",
+    "    bias=True,\n",
+    "    generator=generator,\n",
+    "    teleport_probability=0.2,\n",
+    "    dropout=0.5,\n",
+    "    kernel_regularizer=\"l2\",\n",
+    ")\n",
+    "\n",
+    "x_inp, x_out = appnp.propagate_model(fully_connected_model)\n",
+    "predictions = layers.Softmax()(x_out)\n",
+    "\n",
+    "propagated_model = Model(inputs=x_inp, outputs=predictions)\n",
+    "propagated_model.compile(\n",
+    "    loss=\"categorical_crossentropy\", metrics=[\"acc\"], optimizer=\"Adam\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "1/1 [==============================] - 10s 10s/step - loss: 3.4935 - acc: 0.9051\n",
+      "\n",
+      "Test Set Metrics:\n",
+      "\tloss: 3.4935\n",
+      "\tacc: 0.9051\n"
+     ]
+    }
+   ],
+   "source": [
+    "test_metrics = propagated_model.evaluate(test_gen)\n",
+    "print(\"\\nTest Set Metrics:\")\n",
+    "for name, val in zip(propagated_model.metrics_names, test_metrics):\n",
+    "    print(\"\\t{}: {:0.4f}\".format(name, val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Propagating the MLP with APPNP increases the test set accuracy by ~15% without any further training. As we are performing single-label multiclass classification the accuracy is equivalent to the micro F1 metric. This micro F1 is comparable to that attained in the GraphSAGE paper [2]. GraphSAGE with LSTM aggregation attains a best supervised F1 of 0.95 in [2], however APPNP only required ~3 minutes of training on an 8th gen i7 compared to the hours required for GraphSAGE while still attaining a similar F1."
```

Comment:
I think this is the punchline, maybe make it bold!

Reply:
do you mean **bold** or bold? I can't tell 😆

Reply:
> do you mean **bold** or bold? I can't tell 😆

**bold** 😀